File: leetcode/C++/longest-palindromic-substring.cpp
class Solution {
public:
    int sIdx,eIdx;
    void expand(string &st, int s, int e){
        for(; s>=0 && e<st.length(); --s, ++e){
            if(st[s]!=st[e]) break;
        }
        ++s; --e;
        if(eIdx-sIdx+1<e-s+1) {sIdx=s; eIdx=e;}
    }
    string longestPalindrome(string s) {
        sIdx=eIdx=0;
        for(int i=1; i<s.length(); ++i){
            if(i>1 && s[i-2]==s[i]) expand(s,i-2,i);
            if(s[i]==s[i-1]) expand(s,i-1,i);
        }
        return s.substr(sIdx,eIdx-sIdx+1);
    }
};

File: leetcode/Python/longest-palindromic-substring.py
class Solution:
    def __init__(self):
        self.sIdx = 0
        self.eIdx = 0
    def expand(self, string, s, e):
        while s>=0 and e<len(string):
            if string[s]!=string[e]:
                break
            s-=1
            e+=1
        s+=1
        e-=1
        if self.eIdx-self.sIdx+1 < e-s+1:
            self.sIdx = s
            self.eIdx = e
    def longestPalindrome(self, s: str) -> str:
        for i in range(1,len(s)):
            if s[i]==s[i-1]:
                self.expand(s,i-1,i)
            if i>1 and s[i-2]==s[i]:
                self.expand(s,i-2,i)
        return s[self.sIdx:self.eIdx+1]